### PR TITLE
Allow configuring amount of concurrent tasks

### PR DESCRIPTION
**Pull request by Trial97**
from: https://github.com/PrismLauncher/PrismLauncher/pull/1305#issuecomment-1619129898
> something that would be nice to add to this. the ability to configure / pull the max concurrency from a setting. (ie to configure max concurrent downloads).
> 
> the resent issue about it made be think of this work.

Adds two settings that allow the user to control:
- the number of concurrent tasks
- the number of concurrent network requests

### CODE REVIEW

**Comment by TheKodeToad** [comment]
nice! someone on Discord was asking whether this was possible

**Comment by TheKodeToad** [comment]
I'm guessing there was a smart reason to have some tasks using 6, and others 10.

Maybe you should have a separate network setting?

**Comment by Trial97** [comment]
Should I split that up into two configs then?
Obviously, I also need to fix the tests.

File: launcher/tasks/ConcurrentTask.h [hunk 1]
```diff
@@ -41,14 +41,17 @@
...
 
 class ConcurrentTask : public Task {
     Q_OBJECT
    public:
     using Ptr = shared_qobject_ptr<ConcurrentTask>;
 
-    explicit ConcurrentTask(QObject* parent = nullptr, QString task_name = "", int max_concurrent = 6);
+    explicit ConcurrentTask(QObject* parent = nullptr,
+                            QString task_name = "",
+                            int max_concurrent = APPLICATION->settings()->get("NumberOfConcurrentTasks").toInt());
```
Comment (Ryex) [inline launcher/tasks/ConcurrentTask.h]: This is a bad Idea. Placing a call to `APPLICATION` in a header restricts the header's use to the main exe.
The call should be made and passed to the class where it's needed but eh class default should still be a fixed value

**Review by Ryex — CHANGES_REQUESTED** [review]

**Comment by TheKodeToad** [comment]
> Should I split that up into two configs then? Obviously, I also need to fix the tests.

If you also think it's a good idea :D

File: launcher/tasks/ConcurrentTask.h (hunk 1 above)
Comment (Trial97) [reply launcher/tasks/ConcurrentTask.h]: Yeah I had a similar issue with the test, decided to go with a #ifdef in the two cases the tests were failing

**Review by TheKodeToad — APPROVED** [review]

**Review by DioEgizio — APPROVED** [review]

File: launcher/tasks/ConcurrentTask.h (hunk 1 above)
Comment (TheKodeToad) [reply launcher/tasks/ConcurrentTask.h]: resolved?

Comment (Trial97) [reply launcher/tasks/ConcurrentTask.h]: Partially: I still use applications but in the cpp files. 
The reason I kept this open is that I added some code(a define in cmake) to determine if I should use the setting or not based if we run tests or not.
Also wanted to get Ryes opinion on the current implementation.

**Pull request merged**